Add SCIP Persistent Support

Opt-Mucca opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

## Fixes # .

None.

## Summary/Motivation:

Adds support for SCIP persistent solving. Persistent solving uses the python interface to SCIP directly as opposed to the current standard that involves read and writing a temporary model file.

The new dependency that it introduces is `PySCIPOpt`


## Changes proposed in this PR:
- SCIPPersistent
- SCIPDirect
- Some new tests for the introduced features

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

Opt-Mucca commented:
Two open questions left for this pull request:
- Is the current standard usage of persistent solving through the APPSI? If so then I should consider adding that.
- I avoided tinkering with the GitHub action workflow scripts. I'm not sure if it is appropriate to add a PySCIPOpt dependency there.

For info of how I tested the code:
- `pip install pyscipopt`
- python pyomo/solvers/tests/checks/test_SCIP{Persistent/Direct}.py

mrmundt commented:
@Opt-Mucca - answering your questions:

1. The current standard is APPSI, yes; however, the future standard is going to be using the persistent interface within `pyomo.contrib.solver` (the preview of the reworked solver interface system)
2. If `PySCIPOpt` is needed in order to run tests, then yes, absolutely, it's appropriate. Happy to help with that if you'd like.

Opt-Mucca commented:
@mrmundt Thanks for the fast reply!

1. After having a glance at `pyomo.contrib.solver`, it seems to overlap pretty heavily with the direct interfaces in `pyomo.solvers.plugins`. So that should save on work. I will probably avoid writing the APPSI functionality if there's soon to be a newer interface, but can do if there's demand.
2. Help would be appreciated! I assumed the only changes needed were those that are present in the other PRs #3165 and #2880 (replacing `coptpy` / `maingopy` with `pyscipopt`). If that's right then I can add those lines.

Opt-Mucca commented:
I have now added:
- Some automatic documentation for SCIPPersistent similar to other solver interfaces
- The `pip` instructions to the `.github/workflows`. (I avoided `conda` because I'm not familiar with it. I also only added PySCIPOpt via `pip` when using CPython)

mrmundt commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,838 @@
+
+    def _set_instance(self, model, kwds={}):
+        DirectOrPersistentSolver._set_instance(self, model, kwds)
+        try:
+            self._solver_model = self._scip.Model()
+        except Exception:
+            e = sys.exc_info()[1]
+            msg = (
+                "Unable to create SCIP model. "
+                "Have you installed PySCIPOpt correctly?\n\n\t"
+                + "Error message: {0}".format(e)
+            )
```
It may be preferable to check `self.available` first. If it's not available, then you can raise an error that it couldn't be found. If it's any _other_ error, you can say, "Hey, it's available, but something isn't working correctly. Here is your error message."

Opt-Mucca replied:
I added a call to `self.available()` as suggested. It can't hurt to be there. The call itself raises an appropriate warning by default if the python binding can't be found.

mrmundt commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,838 @@
+        if con.has_lb():
+            if not is_fixed(con.lower):
+                raise ValueError(
+                    "Lower bound of constraint {0} is not constant.".format(con)
```
A general note: This file alone switches between f-strings, `str.format`, and conversion specifiers. It's kind of weird to see all used in a single file.

Opt-Mucca replied:
This is my bad on copying code from the other persistent interfaces, and probably someone else's bad for copying code in turn (I assume that's why there's three standards). 
I am guilty of all the f-strings, and prefer that format, so have changed everything in scip_x.py to match.

mrmundt replied:
Copy-pasta catches us all.  🤷‍♀️

mrmundt commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,838 @@
+            if re.match(suffix, "rc"):
+                extract_reduced_costs = True
+                flag = True
+            if not flag:
+                raise RuntimeError(
+                    "***The scip_direct solver plugin cannot extract solution suffix="
+                    + suffix
+                )
```
Design question: is it better to raise an error or to log (or print loudly) that other requests are being ignored?

(Note: I don't have a strong preference either way.)

Opt-Mucca replied:
I would lean slightly towards the print loudly option, but can see the potential danger for users. As a larger preference though, I would like to stick to similar behavior with the other interfaces.

mrmundt commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,838 @@
+            self.results.solver.status = SolverStatus.aborted
+            self.results.solver.termination_message = (
+                "Optimization terminated because the time expended exceeded "
+                "the value specified in the limits/time parameter."
+            )
+            self.results.solver.termination_condition = (
+                TerminationCondition.maxTimeLimit
+            )
+            soln.status = SolutionStatus.stoppedByLimit
+        elif status == "sollimit" or status == "bestsollimit":
+            self.results.solver.status = SolverStatus.aborted
+            self.results.solver.termination_message = (
+                "Optimization terminated because the number of solutions found "
+                "reached the value specified in the limits/solutions or"
+                "limits/bestsol parameter."
+            )
+            self.results.solver.termination_condition = TerminationCondition.unknown
+            soln.status = SolutionStatus.stoppedByLimit
+        elif status == "memlimit":
+            self.results.solver.status = SolverStatus.aborted
+            self.results.solver.termination_message = (
+                "Optimization terminated because the memory used exceeded "
+                "the value specified in the limits/memory parameter."
+            )
+            self.results.solver.termination_condition = TerminationCondition.unknown
+            soln.status = SolutionStatus.stoppedByLimit
+        elif status == "userinterrupt":
+            self.results.solver.status = SolverStatus.aborted
+            self.results.solver.termination_message = (
+                "Optimization was terminated by the user."
+            )
+            self.results.solver.termination_condition = TerminationCondition.error
+            soln.status = SolutionStatus.error
+        else:
+            self.results.solver.status = SolverStatus.error
+            self.results.solver.termination_message = (
+                "Unhandled SCIP status (" + str(status) + ")"
+            )
+            self.results.solver.termination_condition = TerminationCondition.error
+            soln.status = SolutionStatus.error
```
I recommend putting all of this in its own helper function. It'll clean up `_postsolve`.

Opt-Mucca replied:
Done as suggested

mrmundt commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,838 @@
+
+        self.results.problem.name = scip.getProbName()
+
+        if scip.getObjectiveSense() == "minimize":
+            self.results.problem.sense = minimize
+        elif scip.getObjectiveSense() == "maximize":
+            self.results.problem.sense = maximize
+        else:
+            raise RuntimeError(
+                f"Unrecognized SCIP objective sense: {scip.getObjectiveSense()}"
+            )
```
Do you need this check twice? It also exists in `_set_objective`.

mrmundt commented on pyomo/solvers/plugins/solvers/scip_persistent.py:
```diff
@@ -0,0 +1,190 @@
+
+    def write(self, filename, filetype=""):
+        """
+        Write the model to a file (e.g., and lp file).
```
```suggestion
        Write the model to a file (e.g., an lp file).
```

mrmundt reviewed (changes requested):
Thanks for submitting this PR! Some comments from a quick pass.

mrmundt commented:
@Opt-Mucca - Please make sure you're on the most recent version of `black` and run (in the root): `black . -S -C  --exclude examples/pyomobook/python-ch/BadIndent.py`

Opt-Mucca commented on .github/workflows/test_branches.yml:
```diff
@@ -268,6 +268,12 @@ jobs:
+            if [[ ${{matrix.python}} == pypy* ]]; then
+                echo "skipping SCIP for pypy"
+            else
+                python -m pip install --cache-dir cache/pip pyscipopt==5.0.0 \
```
Removed the specific version. I added this as it will fail for any `pyscipopt<=4.3.0`.
Added the appropriate conda line.

Opt-Mucca commented on .github/workflows/test_pr_and_main.yml:
```diff
@@ -298,6 +298,12 @@ jobs:
+            if [[ ${{matrix.python}} == pypy* ]]; then
+                echo "skipping SCIP for pypy"
+            else
+                python -m pip install --cache-dir cache/pip pyscipopt==5.0.0 \
```
Same as above

Opt-Mucca commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,838 @@
+                soln_variables = soln.variable
+                soln_constraints = soln.constraint
+                scip_sol = scip.getBestSol()
+
+                scip_vars = scip.getVars()
+                scip_var_names = [scip_var.name for scip_var in scip_vars]
+                var_names = set(self._solver_var_to_pyomo_var_map.keys())
+                assert set(scip_var_names) == var_names
+                var_vals = [scip.getVal(scip_var) for scip_var in scip_vars]
+
+                for scip_var, val, name in zip(scip_vars, var_vals, scip_var_names):
+                    pyomo_var = self._solver_var_to_pyomo_var_map[name]
+                    if self._referenced_variables[pyomo_var] > 0:
+                        soln_variables[name] = {"Value": val}
+
+                if extract_reduced_costs:
+                    vals = [scip.getVarRedcost(scip_var) for scip_var in scip_vars]
+                    for scip_var, val, name in zip(scip_vars, vals, scip_var_names):
+                        pyomo_var = self._solver_var_to_pyomo_var_map[name]
+                        if self._referenced_variables[pyomo_var] > 0:
+                            soln_variables[name]["Rc"] = val
+
+                if extract_duals or extract_slacks:
+                    scip_cons = scip.getConss()
+                    con_names = [cons.name for cons in scip_cons]
+                    assert set(self._solver_con_to_pyomo_con_map.keys()) == set(
+                        con_names
+                    )
+                    for name in con_names:
+                        soln_constraints[name] = {}
+
+                if extract_duals:
+                    vals = [scip.getDualSolVal(con) for con in scip_cons]
+                    for val, name in zip(vals, con_names):
+                        soln_constraints[name]["Dual"] = val
+
+                if extract_slacks:
+                    vals = [scip.getSlack(con, scip_sol) for con in scip_cons]
+                    for val, name in zip(vals, con_names):
+                        soln_constraints[name]["Slack"] = val
```
I decided against this suggestion. I'm not a fan of helper functions with a huge amount of arguments, and I think this code isn't as overly repetitive as the status handling.

mrmundt replied:
Makes sense to me

Opt-Mucca commented on pyomo/solvers/plugins/solvers/scip_persistent.py:
```diff
@@ -0,0 +1,190 @@
+        ----------
+        param: str
+            The SCIP parameter to set. Options include any SCIP parameter.
+            Please see the SCIP documentation for options.
```
Nice suggestion! Added a link

Opt-Mucca commented on pyomo/solvers/plugins/solvers/scip_persistent.py:
```diff
@@ -0,0 +1,190 @@
+        self._solver_model.setObjective(obj_coef * scip_var, sense=sense, clear=False)
+
+    def reset(self):
+        """ This function is necessary to call before making any changes to the
```
Thank you! I'm hoping that it's clear enough, and that this won't cause a headache for many users.

Opt-Mucca commented:
@mrmundt Black should be run and all the issues responded to.

mrmundt commented:
@Opt-Mucca - Spelling is the new failure. (You are suffering the same issue as all the core devs... The auto-linter and spell checker.)

Opt-Mucca commented:
@mrmundt TIL there are auto-linters that double as spell checkers. (I'll fix the spelling errors that come up each time the test fails. If it's just the two errors then I'm properly amazed.)

mrmundt commented:
@Opt-Mucca - If you could have heard the griping in our weekly dev call when I introduced the spell checker...

Opt-Mucca commented:
@mrmundt I can imagine (would probably be guilty of griping even if makes sense)

For the currently failing runs: I've added a fix that should work, but I will ask some colleagues tomorrow. (I'm not the greatest the warm start code). The fixes raise some warning where a loaded solution has variables of incorrect values (this may stem from non specified variables having a default value of 0)

The current reason of the pipeline failure: SCIP was trying to load an invalid solution before checking for feasibility (I think this is a longstanding python error with the interface and one of the functions)

The added solution: I simply added a check on the feasibility of the solution before adding it to the solution storage.

I will check with colleagues, but I'm not sure on how SCIP handles partial solutions. I am not familiar or don't know if we have any functionality like Gurobi that tries to complete the solution. This also holds for trying to repair a given infeasible solution.

Opt-Mucca commented:
Support for partial solution loading is now added. The failed tests are now passing locally too.

Opt-Mucca commented:
@mrmundt I am completely unsure on the protocol here. I've never done a merge request where I was not a member of the repository. 

The failed test now passes locally, but I am unable to trigger any pipelines here.

mrmundt commented:
@Opt-Mucca - Please make sure to read our [Contribution Guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) as it'll answer most questions about our process.

As for triggering tests here, because you are a first-time contributor, one of our core dev team members needs to approve them (which I have been doing for this PR).

Opt-Mucca commented:
@mrmundt This will take awhile.
- I fixed the current issue (I was accessing the transformed problem after solving and the pyomo map was failing + SOS and nonlinear constraints can't access slack information). Such a fix required some new wrapped functions in PySCIPOpt, so will have to add them and release a new version.
- There are still errors in `test_writers`. I don't see this being easy. Most all (aside from an error with trivial constraints) stem from the reduced costs + slack being incorrect. This information is always weird to handle with SCIP for non-LP problems, so will need to think on it.

Opt-Mucca commented:
@mrmundt After talking to some colleagues we decided to not support reduced cost and dual values. They are simply too difficult to access for SCIP in a context that people would understand without disabling presolve (The loss in performance just doesn't outweigh the people that want to access this information).

There are currently two issues, both of which I would appreciate some advice or help on:
- I cannot calculate the slack correctly. After some tricks with ranged constraints I've ensured that SCIP is outputting the correct values, but the baseline comparisons seem to arbitrarily take either `rhs - activity` or `activity - lhs`. Internally we take the minimum of the two, but that doesn't seem to be the cases in the tests. Defaulting to one side in the case of ranged constraints also seems to fail. 
- I am not sure how to handle trivial constraints. I have set the flag `_skip_trivial_constraints=True`, but the corresponding tests are now failing. Adding a dummy constraint key to the internal dictionary mappings also doesn't seem to help.

blnicho commented:
We discussed this during the developer call and would advise that you not support the `slack` suffix. This is old functionality that we don't intend to support in the new solver interfaces and the same information can be calculated using a simple utility function. We're supportive if you want to skip trivial constraints, however this will likely require some updates to baseline tests.

Opt-Mucca commented:
I have now removed support for the `slack` suffix. 

I am still a bit unsure on what to do with the trivial constraints. The Python interface for SCIP just simply doesn't allow dummy constraints to be added. Some variables have to feature somewhere in the constraint. 
I think changing the tests to skip solvers with the `_skip_trivial_constraints=True` flag is a good idea. The alternative would be to pretend there is a constraint on the SCIP interface side, which would be extremely messy. Is anyone willing to change the test scope on the Pyomo dev team side?

mrmundt commented:
@Opt-Mucca - We have been discussing redesigning the solver test suite for a fair bit of time and have added it as an action item to #1030 . We will address this as a corner case for that.

Opt-Mucca commented:
@mrmundt Can you please run the pipeline now? I put `scip_persistent` as an expected fail for the `LP_trivial_constraint` test case. Everything is now passing locally.

Opt-Mucca commented:
Hmmmmmmmmmmmmmm. I didn't have scipy installed when I tested locally so it skipped that test. The error is that `np.float` is a non-allowed type when using `PySCIPOpt`. The user must give Python `float` or `int` types. Should probably consider expanding on that. For now there's two options, both of which are pretty fast:
- I just check if the user is giving non-standard Python values, e.g. `numpy.float64`, and convert them to Python (Will do this unless you advise against). Downside is this adds a tiny bit of overhead.
- Skip tests where `numpy.float64` values are passed as LHS and RHS values.

mrmundt commented:
@Opt-Mucca - Our general thought process is "protect users from themselves." I would recommend converting them but also log / alert the user so they know the conversion is happening.

Opt-Mucca commented:
@mrmundt I've added a warning now for each time a constraint is added and the RHS / LHS needs to be changed.
I had to use `type(x) == y`, because `isinstance(x, float)` also passes for `np.float` (didn't know that). The failed tests are now passing.

Opt-Mucca commented:
@mrmundt I'm not sure if the failing tests are because of my changes. Can I get a second opinion on that?

Opt-Mucca commented:
The latest change is pretty minor, but would mean that it's not backwards compatible to `pyscipopt` versions from years ago (I think compatibility would now be the last 2-3 releases). There was some basic missing functionality for retrieving the original constraints after starting / ending the solving process, and the Pyomo implementation was then giving some incorrect statistics.

mrmundt commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,769 @@
+        if obj.active is False:
+            raise ValueError("Cannot add inactive objective to solver.")
+
+        if obj.sense == minimize:
+            sense = "minimize"
+        elif obj.sense == maximize:
+            sense = "maximize"
+        else:
+            raise ValueError(f"Objective sense is not recognized: {obj.sense}")
```
I would instead do:
```
if obj.sense not in ['maximize', 'minimize']:
    raise ValueError(f"Objective sense is not recognized: {obj.sense}")
...
self._solver_model.setObjective(scip_expr, sense=obj.sense)
```

Opt-Mucca replied:
This does not work because you're comparing to string objects. `maximize` and `minimzie` in the code above are imports from pyomo.

I think I could change it to `obj.sense.value` or `obj.sense.name` (can't remember), but I wasn't 100% confident in covering all cases. I've left it as is.

mrmundt commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,769 @@
+                soln.gap = None
+
+        self.results.problem.number_of_constraints = scip.getNConss(transformed=False)
+        # self.results.problem.number_of_nonzeros = None
```
Remove?

Opt-Mucca replied:
Removed

mrmundt commented on pyomo/solvers/tests/checks/test_SCIPDirect.py:
```diff
@@ -0,0 +1,335 @@
+)
+from pyomo.opt import SolverFactory, TerminationCondition, SolutionStatus
+
+try:
+    import pyscipopt
+
+    scip_available = True
+except ImportError:
+    scip_available = False
```
I will ping @jsiirola again about this thought.

mrmundt reviewed (approved):
Sorry, I meant to click "Approve" - whoops. I left a few comments but none of them are so substantial that they should hold up this PR. They are mostly stylistic.

mrmundt commented:
FYI, the doc build/doc test failures are NOT RELATED. I have #3479 open to resolve it.

Opt-Mucca commented:
I added a minor change: I no longer check what type the `LHS` of a constraint is. I added it originally because `pyomo` supports `numpy` floats and ints being passed, and I observed some crashes in SCIP's model creation. I looked into this a month or two ago, and concluded that it only happens from `np.version() < 2` and for `LHS` values. So I've just removed the check for the RHS. This should save many users getting spammed with warnings. Associated issues if anyone wants to help: https://github.com/scipopt/PySCIPOpt/issues/937

mrmundt commented:
@Opt-Mucca - can you do me one last favor please? I updated all of the existing files recently to reflect that our copyright is 2008 - 2025 now (see PR #3515 ) - could you please update your new files here to reflect that as well?

jsiirola commented on pyomo/solvers/tests/checks/test_SCIPPersistent.py:
```diff
@@ -0,0 +1,318 @@
+)
+from pyomo.opt import SolverFactory
+
+try:
+    import pyscipopt
+
+    scip_available = True
+except ImportError:
+    scip_available = False
```
In general, yes - but as this is in a testing file, it is not a big deal.

jsiirola reviewed (changes requested):
Apologies for how long it took to get to this.  Overall, this is not in too bad of shape, but there are a couple things that would be very good to rework:
- avoid the use of StandardRepn, and instead use / write a walker based on `StreamBasedExpressionVisitor`.
- rework this so we only need to walk each expression once (this is a performance issue, and is particularly important for persistent interfaces where performance is one of the big motivators).

Opt-Mucca commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,762 @@
+
+        new_expr = repn.constant
+
+        if len(repn.linear_vars) > 0:
+            referenced_vars.update(repn.linear_vars)
+            new_expr += sum(
+                repn.linear_coefs[i] * self._pyomo_var_to_solver_var_expr_map[var]
+                for i, var in enumerate(repn.linear_vars)
+            )
+
+        for i, v in enumerate(repn.quadratic_vars):
+            x, y = v
+            new_expr += (
+                repn.quadratic_coefs[i]
+                * self._pyomo_var_to_solver_var_expr_map[x]
+                * self._pyomo_var_to_solver_var_expr_map[y]
+            )
+            referenced_vars.add(x)
+            referenced_vars.add(y)
```
Hey @jsiirola thanks for the review! It's now my turn to apologise for taking so long though...... Have started a new job and simply been waiting on a burst of motivation some evening to look at this.

I used the representation compiler because it was the first thing that I found and seemed to fit what I wanted. I do not have an overview whatsoever of the various Pyomo data structures, so that's on me for not doing enough research. I will try and look into `StreamBasedExpressionVisitor` and make the relevant changes (No complaints from me if someone else on the team wants to do the changes themself). Is there some particular date for when it would be good if this PR was finished?

Opt-Mucca commented on pyomo/solvers/plugins/solvers/scip_direct.py:
```diff
@@ -0,0 +1,762 @@
+        if (max_degree is not None) and (degree > max_degree):
+            raise DegreeError(
+                "While SCIP supports general non-linear constraints, the objective must be linear. "
+                "Please reformulate the objective by introducing a new variable. "
```
I am on of the opinion that the modelling language should not be doing something that the user didn't specifically ask for.  For example, creating some "shadow" variable that is likely never revealed to the user and would cause a mismatch in number of variables in the model and the number of variables in the users list / dict of "standard" variables. We have had many debates about this in `PySCIPOpt`.
Completely fair to have the opposite opinion of "Make it as easy as possible for the user"

jsiirola replied:
In general, I completely agree.  However, we have historically made exceptions for trivial changes that 
- can only be implemented in a single way
- enable a solver to be run on a model that a rational user would expect the solver to run on (that is, we want users to be able to easily switch between solvers without having to recode their model)

Examples of changes that we do make are:
- Adding a trivial (fixed) variable to LP files so that we can add constant terms to objectives (several solvers do not accept a constant as part of the objective expression in LP format)
- Converting the optimization sense (max to min) for solvers that only accept one sense.

In my opinion, this would fall into the same category...  But I also think that this is on the edge, and tossing a warning for the user would be completely warranted.  (I can also be convinced that this reformulation is a "bridge too far" and the error is the right path forward).

michaelbynum commented:
@Opt-Mucca, we are refactoring the solver interfaces in pyomo/contrib/solver (https://pyomo.readthedocs.io/en/stable/explanation/experimental/solvers.html). We would really like this functionality there instead of in pyomo/solvers/plugins/solvers (which will eventually become deprecated). I plan to build off your branch by moving the files into pyomo/contrib/solver and reworking them to match the new interfaces. Please let us know if you have any objections/concerns.

Opt-Mucca commented:
@michaelbynum No complaints at all. Sorry in advance if the code is a bit sloppy (mainly the SCIP naming and constant checks there's no `numpy.float` as a LHS). Feel free to ping me if you have questions about it.
Does that mean that there's no pressure to complete this PR before mid August?
